From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sumit Gupta <sumitg@nvidia.com>
Date: Wed, 7 Sep 2022 18:42:17 +0530
Subject: [PATCH 175/240] NVIDIA: SAUCE: soc: tegra: cbb: add multi socket
 error handling

Add support to handle errors from multiple sockets by using
the per socket interrupts for each CBB fabric.
This is done by affining the CBB error notifer interrupt of a
fabric to the CPU in local Socket.

Bug 4121200

Change-Id: Iae55dcf1c81c8044f245a011612b9d88212d2be0
Signed-off-by: Sumit Gupta <sumitg@nvidia.com>

diff --git a/drivers/soc/tegra/cbb/tegra234-cbb.c b/drivers/soc/tegra/cbb/tegra234-cbb.c
index fb26f085f6914..9215509967234 100644
--- a/drivers/soc/tegra/cbb/tegra234-cbb.c
+++ b/drivers/soc/tegra/cbb/tegra234-cbb.c
@@ -13,6 +13,7 @@
 #include <linux/acpi.h>
 #include <linux/clk.h>
 #include <linux/cpufeature.h>
+#include <linux/cpuhotplug.h>
 #include <linux/debugfs.h>
 #include <linux/module.h>
 #include <linux/of.h>
@@ -181,6 +182,9 @@ struct tegra234_cbb {
 	u32 mn_attr1;
 	u32 mn_attr2;
 	u32 mn_user_bits;
+
+	int numa_node;
+	struct hlist_node cpuhp_node;
 };
 
 static inline struct tegra234_cbb *to_tegra234_cbb(struct tegra_cbb *cbb)
@@ -229,6 +233,8 @@ tegra234_cbb_write_access_allowed(struct platform_device *pdev, struct tegra234_
 	return false;
 }
 
+static int tegra234_cbb_hp_state;
+
 static void tegra234_cbb_fault_enable(struct tegra_cbb *cbb)
 {
 	struct tegra234_cbb *priv = to_tegra234_cbb(cbb);
@@ -733,8 +739,14 @@ static int tegra234_cbb_interrupt_enable(struct tegra_cbb *cbb)
 	struct tegra234_cbb *priv = to_tegra234_cbb(cbb);
 
 	if (priv->sec_irq) {
-		int err = devm_request_irq(cbb->dev, priv->sec_irq, tegra234_cbb_isr, 0,
-					   dev_name(cbb->dev), priv);
+		unsigned long irqflags = 0;
+		int err = 0;
+
+		if (num_possible_nodes() > 1)
+			irqflags = IRQF_NOBALANCING;
+
+		err = devm_request_irq(cbb->dev, priv->sec_irq, tegra234_cbb_isr,
+				       irqflags, dev_name(cbb->dev), priv);
 		if (err) {
 			dev_err(cbb->dev, "failed to register interrupt %u: %d\n", priv->sec_irq,
 				err);
@@ -1681,6 +1693,49 @@ static const struct acpi_device_id tegra241_cbb_acpi_ids[] = {
 };
 MODULE_DEVICE_TABLE(acpi, tegra241_cbb_acpi_ids);
 
+static int tegra234_cbb_online_cpu(unsigned int cpu, struct hlist_node *cpuhp_node)
+{
+	struct tegra234_cbb *cbb;
+	struct cpumask mask;
+
+	cbb = hlist_entry_safe(cpuhp_node, struct tegra234_cbb, cpuhp_node);
+
+	if (num_possible_nodes() > 1 && cbb->numa_node == cpu_to_node(cpu)) {
+		cpumask_clear(&mask);
+		cpumask_and(&mask, cpumask_of_node(cbb->numa_node), cpu_online_mask);
+
+		irq_set_affinity(cbb->sec_irq, &mask);
+
+		pr_debug("%s, cpu:%d, fab:%s, numa_node:0x%x, cpu_to_node:%d, mask:%*pbl\n",
+			 __func__, cpu, cbb->fabric->fab_list[cbb->fabric->fab_id].name,
+			 cbb->numa_node, cpu_to_node(cpu), cpumask_pr_args(&mask));
+	}
+
+	return 0;
+}
+
+static int tegra234_cbb_offline_cpu(unsigned int cpu, struct hlist_node *cpuhp_node)
+{
+	struct tegra234_cbb *cbb;
+	struct cpumask mask;
+
+	cbb = hlist_entry_safe(cpuhp_node, struct tegra234_cbb, cpuhp_node);
+
+	if (num_possible_nodes() > 1 && cbb->numa_node == cpu_to_node(cpu)) {
+		cpumask_clear(&mask);
+		cpumask_and(&mask, cpumask_of_node(cbb->numa_node), cpu_online_mask);
+		cpumask_andnot(&mask, &mask, cpumask_of(cpu));
+
+		irq_set_affinity(cbb->sec_irq, &mask);
+
+		pr_debug("%s, cpu:%d, fab:%s, numa_node:0x%x, cpu_to_node:%d, mask:%*pbl\n",
+				__func__, cpu, cbb->fabric->fab_list[cbb->fabric->fab_id].name,
+				cbb->numa_node, cpu_to_node(cpu), cpumask_pr_args(&mask));
+	}
+
+	return 0;
+}
+
 static int tegra234_cbb_probe(struct platform_device *pdev)
 {
 	const struct tegra234_cbb_fabric *fabric;
@@ -1700,6 +1755,7 @@ static int tegra234_cbb_probe(struct platform_device *pdev)
 			dev_err(&pdev->dev, "no device match found\n");
 			return -ENODEV;
 		}
+		set_dev_node(&pdev->dev, acpi_get_node(ACPI_HANDLE(&pdev->dev)));
 	}
 
 	cbb = devm_kzalloc(&pdev->dev, sizeof(*cbb), GFP_KERNEL);
@@ -1730,15 +1786,44 @@ static int tegra234_cbb_probe(struct platform_device *pdev)
 		return 0;
 	}
 
+	/* set ERD bit to mask SError and generate interrupt to report error */
+	if (cbb->fabric->off_mask_erd)
+		tegra234_cbb_mask_serror(cbb);
+
+	cbb->numa_node = dev_to_node(&pdev->dev);
+	if (cbb->numa_node != NUMA_NO_NODE && num_possible_nodes() > 1) {
+		dev_dbg(&pdev->dev, "fabric:%s, cbb->numa_node:0x%x\n",
+			cbb->fabric->fab_list[cbb->fabric->fab_id].name, cbb->numa_node);
+
+		err = cpuhp_state_add_instance(tegra234_cbb_hp_state, &cbb->cpuhp_node);
+		if (err)
+			return err;
+	}
+
+	err = tegra_cbb_register(&cbb->base);
+	if (err)
+		goto remove_hp_inst;
+
 	spin_lock_irqsave(&cbb_lock, flags);
 	list_add(&cbb->base.node, &cbb_list);
 	spin_unlock_irqrestore(&cbb_lock, flags);
 
-	/* set ERD bit to mask SError and generate interrupt to report error */
-	if (cbb->fabric->off_mask_erd)
-		tegra234_cbb_mask_serror(cbb);
+remove_hp_inst:
+	if (err && (tegra234_cbb_hp_state > 0))
+		cpuhp_state_remove_instance_nocalls(tegra234_cbb_hp_state, &cbb->cpuhp_node);
+
+	return err;
+}
+
+static void tegra234_cbb_remove(struct platform_device *pdev)
+{
+	struct tegra234_cbb *cbb = platform_get_drvdata(pdev);
+
+	if (!cbb)
+		return;
 
-	return tegra_cbb_register(&cbb->base);
+	if (tegra234_cbb_hp_state > 0)
+		cpuhp_state_remove_instance_nocalls(tegra234_cbb_hp_state, &cbb->cpuhp_node);
 }
 
 static int __maybe_unused tegra234_cbb_resume_noirq(struct device *dev)
@@ -1762,6 +1847,7 @@ static const struct dev_pm_ops tegra234_cbb_pm = {
 
 static struct platform_driver tegra234_cbb_driver = {
 	.probe = tegra234_cbb_probe,
+	.remove = tegra234_cbb_remove,
 	.driver = {
 		.name = "tegra234-cbb",
 		.of_match_table = tegra234_cbb_dt_ids,
@@ -1772,12 +1858,30 @@ static struct platform_driver tegra234_cbb_driver = {
 
 static int __init tegra234_cbb_init(void)
 {
-	return platform_driver_register(&tegra234_cbb_driver);
+	int ret;
+
+	if (num_possible_nodes() > 1) {
+		ret = cpuhp_setup_state_multi(CPUHP_AP_ONLINE_DYN, "tegra234_cbb:online",
+					      tegra234_cbb_online_cpu, tegra234_cbb_offline_cpu);
+		if (ret < 0)
+			return ret;
+
+		tegra234_cbb_hp_state = ret;
+	}
+
+	ret = platform_driver_register(&tegra234_cbb_driver);
+	if (ret && (tegra234_cbb_hp_state > 0))
+		cpuhp_remove_multi_state(tegra234_cbb_hp_state);
+
+	return ret;
+
 }
 pure_initcall(tegra234_cbb_init);
 
 static void __exit tegra234_cbb_exit(void)
 {
+	if (tegra234_cbb_hp_state > 0)
+		cpuhp_remove_multi_state(tegra234_cbb_hp_state);
 	platform_driver_unregister(&tegra234_cbb_driver);
 }
 module_exit(tegra234_cbb_exit);
-- 
2.43.0

